### PR TITLE
Derive Service Worker scope from path

### PR DESCRIPTION
**Pull request by k-antipochkin**
### Description of Changes

Derive Service Worker scope from path and remove requirement to control the scope.

This enables hosting the Service Worker file at a non-root location, or using multiple Service Workers.

### Breaking Changes

-   None

### Release Checklist

Prepare:

-   ~~[ ] Detail any breaking changes. Breaking changes require a new major version number~~

Bump versions in:

-  ~~[ ] package.json~~
-   ~~[ ] package-lock.json (run `npm install` after editing package.json)~~
-   [x] src/core/index.ts

Build:

-   [x] `npm run dist`
-   [x] Commit resulting `dist/` changes

Release:

-   ~~[ ] Squash and merge to master~~
-   ~~[ ] Delete branch once merged~~
-   ~~[ ] Create tag from master matching chosen version~~
-  ~~[ ] Fill out release notes~~
-   ~~[ ] Run `npm publish --access public`~~
-   [ ] Publish `dist/worker.js` and `dist/web-bundle.js` to CDN (re-run `npm run dist` locally to update `web-bundle.js`)

### CODE REVIEW

File: src/core/utils.ts [hunk 1]
```diff
@@ -173,18 +173,18 @@ export function base64UrlEncode(buffer: ArrayBuffer): string {
     return urlVariant;
 }
 
-export function registerServiceWorker(
-    path: string
-): Promise<ServiceWorkerRegistration> {
+export async function registerServiceWorker(workerPath: string){
```
Comment (cgwyllie) [inline src/core/utils.ts:176]: @k-antipochkin can you check tooling for Prettier? looks like formatting hasn't been applied on save in these files. Thanks!

Comment (k-antipochkin) [reply src/core/utils.ts]: Formatted the file, those lines remained the same though

File: src/index.ts [hunk 2]
```diff
@@ -100,6 +82,29 @@ export default class Kumulos {
         }
     }
 
+    private initializePushFeature() {
+        trackOpenFromQuery(this.context);
+        registerServiceWorker(this.context.serviceWorkerPath);
+
+        new PromptManager(
```
Comment (cgwyllie) [inline src/index.ts]: Is there any risk of GC of the manager now we don't assign it to a property?

File: src/index.ts [hunk 3]
```diff
@@ -100,6 +82,29 @@ export default class Kumulos {
...
+            this.rootFrame
+        );
+
+        this.maybeAddMessageEventListenerToSW();
+        this.maybeFireOpenedHandler();
+    }
+
+    private maybeAddMessageEventListenerToSW() {
+        if ('serviceWorker' in navigator) {
+            navigator.serviceWorker.addEventListener(
```
Comment (cgwyllie) [inline src/index.ts:97]: Now this is refactored, maybe we can invert condition & early return to reduce nesting of the block?

File: src/core/utils.ts [hunk 4]
```diff
@@ -173,18 +173,18 @@ export function base64UrlEncode(buffer: ArrayBuffer): string {
...
-            'ServiceWorker is not supported in this browser, aborting...'
-        );
+        console.error('ServiceWorker is not supported in this browser, aborting...');
+        return;
     }
 
-    return navigator.serviceWorker.register(path).then(() => {
-        return navigator.serviceWorker.ready;
-    });
+    const updatedPath = new URL(workerPath, location.origin).href;
```
Comment (cgwyllie) [inline src/core/utils.ts]: Any value in extracting the building of full SW URL to a util? (minor duplication in other places)

File: src/index.ts (hunk 2 above)
Comment (k-antipochkin) [reply src/index.ts]: Shouldnt be a problem as the register is done on navigator.serviceWorker (navigator references the service worker). The registration is accessed by navigator.serviceWorker then. Dont see a reason to keep a reference to the registration.

File: src/index.ts (hunk 3 above)
Comment (k-antipochkin) [reply src/index.ts]: Done

File: src/core/utils.ts (hunk 4 above)
Comment (k-antipochkin) [reply src/core/utils.ts]: Done

File: src/core/utils.ts (hunk 1 above)
Comment (cgwyllie) [reply src/core/utils.ts]: Cool, think some spacings tweaked which is good, yeah it doesn't always line break in prettier unless over some soft wrap I think.

File: src/index.ts (hunk 2 above)
Comment (cgwyllie) [reply src/index.ts]: Sorry not reference to the registration, but the PromptManager itself is responsible for drawing soft prompts for web opt in, if it is GC maybe it won't run logic to show the UI elements or handle button presses? (it's possible there's some internal ref or cycle, but just cautious here).

Have we regression tested prompts with this build?

Comment (k-antipochkin) [reply src/index.ts]: Oh I see. Regarding the PromptManager, there were no references to the object from outside. The object shouldn't be destroyed unless there is no "potential" code anymore. So as long as we are awaiting something inside / create another object and pass callbacks (onPromptAccepted={this.onPromptAccepted} for example) the object will stay. It should be cleared when its done execution and no one references it. In general, I don't think that its a good idea when the constructor is the only exposed method but lets refactor this later.

**Pull request merged**